build-tools/packages/build-cli/src/library/git.ts

```diff
@@ -251,4 +294,142 @@ export class Repository {
 		// Files are already repo root-relative
 		return [...allFiles];
 	}
+
+	/**
+	 * Map of release group/package name to all the git tags for the group/package.
+	 */
+	private readonly _tags: Map<string, string[]> = new Map();
+
+	/**
+	 * Returns an array of all the git tags associated with a release group.
+	 *
+	 * @param releaseGroupOrPackage - The release group or independent package to get tags for.
+	 * @returns An array of all all the tags for the release group or package.
+	 */
+	public async getTagsForReleaseGroup(releaseGroupOrPackage: string): Promise<string[]> {
+		// eslint-disable-next-line import/no-deprecated
+		const prefix = isMonoRepoKind(releaseGroupOrPackage)
+			? releaseGroupOrPackage.toLowerCase()
+			: PackageName.getUnscopedName(releaseGroupOrPackage);
+		const cacheEntry = this._tags.get(prefix);
+		if (cacheEntry !== undefined) {
+			return cacheEntry;
+		}
+
+		const tags = await this.gitClient.tags(["--list", `${prefix}_v*`]);
+		this._tags.set(prefix, tags.all);
+		return tags.all;
+	}
+
+	private readonly _versions: Map<string, VersionDetails[]> = new Map();
+
+	/**
+	 * Gets all the versions for a release group or independent package. This function only considers the tags in the
+	 * repo to determine releases and dates.
+	 *
+	 * @param releaseGroupOrPackage - The release group or independent package to get versions for.
+	 * @returns An array of {@link ReleaseDetails} containing the version and date for each version.
+	 */
+	public async getAllVersions(
+		releaseGroupOrPackage: string,
+	): Promise<VersionDetails[] | undefined> {
+		const cacheEntry = this._versions.get(releaseGroupOrPackage);
+		if (cacheEntry !== undefined) {
+			return cacheEntry;
+		}
+
+		const versions = new Map<string, Date>();
+		const tags = await this.getTagsForReleaseGroup(releaseGroupOrPackage);
+
+		for (const tag of tags) {
+			const ver = getVersionFromTag(tag);
+			if (ver !== undefined && ver !== "" && ver !== null) {
+				// eslint-disable-next-line no-await-in-loop
+				const rawDate = await this.gitClient.show([
+					// Suppress the diff output
+					"-s",
+					// ISO 8601 format
+					"--format=%cI",
+					// the git ref - a tag in this case
+					tag,
+				]);
+				const date = parseISO(rawDate.trim());
+				versions.set(ver, date);
+			}
+		}
+
+		if (versions.size === 0) {
+			return undefined;
+		}
+
+		const toReturn: VersionDetails[] = [];
+		for (const [version, date] of versions) {
+			toReturn.push({ version, date });
+		}
+
+		this._versions.set(releaseGroupOrPackage, toReturn);
+		return toReturn;
+	}
+
+	/**
+	 * Create a branch with name. throw an error if the branch already exists.
+	 */
+	public async createBranch(branchName: string): Promise<void> {
+		if (await this.getShaForBranch(branchName)) {
+			throw new Error(`${branchName} already exists. Failed to create.`);
+		}
+		await this.createBranch(branchName);
```

Comment:
Is this not going into an infinite recursion calling itself?

Reply:
Good catch.